bril-ts/brili.ts

```diff
@@ -708,12 +749,84 @@ function evalInstr(instr: bril.Instruction, state: State): Action {
     return {"action": "commit"};
   }
 
+  case "pack": {
+    let values: Value[] = [];
+    for (let i = 0; i < (instr.args || []).length; i++) {
+      values.push(getArgument(instr, state.env, i));
+    }
+    state.env.set(instr.dest, values);
+    return NEXT;
+  }
+
+  case "unpack": {
+    if (!instr.args || instr.args.length < 2) {
+      throw error("malformed unpack instr");
+    }
+    let tuple = getArgument(instr, state.env, 0);
+    let index = parseInt(instr.args[1]);
+    if (!Array.isArray(tuple)) {
+      throw error("cannot unpack a non-tuple value");
+    }
+    if (tuple.length <= index) {
+      throw error("invalid tuple index: " + index +
+                  ", length: " + tuple.length);
+    }
+    state.env.set(instr.dest, tuple[index]);
+    return NEXT;
+  }
+
+  case "construct": {
+    if (!instr.args || instr.args.length < 2) {
+      throw error("malformed construct instr");
+    }
+    if (typeof(instr.type) !== "object" || !instr.type.hasOwnProperty("sum")) {
+      throw error("construct destination must be a sum type");
+    }
+    let value = getArgument(instr, state.env, 0);
+    let index = parseInt(instr.args[1]);
```

Comment:
TBH, I am still a little concerned about this syntax change. I don't have an alternative to propose, unfortunately, but it is pretty weird that _for every other instruction_, arguments are variable names, but for _this one special instruction_, the argument is a string that encodes a decimal integer and not a variable name at all. For example, consider writing a def/use analysis: you would now have to say "all the things in `args` are variable uses, *except for this one instruction case*, where one of the arguments does *not* name a variable."

Like I said, not sure what to do about this, but I think we may want to reconsider the abstract syntax for this instruction…